src/iommu/iommufd.c

```diff
@@ -312,6 +329,66 @@ static int iommufd_open(void)
 	return 0;
 }
 
+#ifdef HAVE_IOMMU_FAULT_QUEUE_ALLOC
+int iommufd_alloc_fault_queue(struct iommufd_fault_queue *fq)
+{
+	struct iommu_fault_alloc fault_alloc_cmd = {
+		.size = sizeof(fault_alloc_cmd),
+	};
+
+	if (ioctl(__iommufd, IOMMU_FAULT_QUEUE_ALLOC, &fault_alloc_cmd)) {
+		log_debug("could not allocate fault queue\n");
+		return -1;
+	}
+
+	*fq = (struct iommufd_fault_queue) {
+		.fault_id = fault_alloc_cmd.out_fault_id,
+		.fault_fd = fault_alloc_cmd.out_fault_fd,
+	};
+
+	return 0;
+}
+
+int iommufd_set_fault_queue(struct iommu_ctx *ctx, struct iommufd_fault_queue *fq, int devfd)
+{
+	struct iommu_ioas *ioas = container_of_var(ctx, ioas, ctx);
+	struct iommufd_device_info *info;
+
+	struct iommu_hwpt_alloc hwpt_alloc_cmd = {
+		.size = sizeof(hwpt_alloc_cmd),
+		.flags = IOMMU_HWPT_FAULT_ID_VALID,
+		.pt_id = ioas->id,
+		.fault_id = fq->fault_id,
+	};
+
+	struct vfio_device_attach_iommufd_pt attach_cmd = {
+		.argsz = sizeof(attach_cmd),
+	};
+
+	info = iommufd_get_device_info(devfd);
+	if (!info)
+		return -1;
+
+	hwpt_alloc_cmd.dev_id = info->dev_id;
+
+	if (ioctl(__iommufd, IOMMU_HWPT_ALLOC, &hwpt_alloc_cmd)) {
+		log_debug("could not allocate hardware page table\n");
+		return -1;
+	}
+
+	info->hwpt_id = hwpt_alloc_cmd.out_hwpt_id;
+
+	attach_cmd.pt_id = hwpt_alloc_cmd.out_hwpt_id;
+
+	if (ioctl(devfd, VFIO_DEVICE_ATTACH_IOMMUFD_PT, &attach_cmd)) {
```

Comment:
## What about `VFIO_DEVICE_DETACH_IOMMUFD_PT`

At the moment the way to use the IOPF functionality in libvfn is by using a
`VFIO_DEVICE_ATTACH_IOMMUFD_PT` ioctl and then it is detached when the pci
device is closed. Couple of questions here:
1. Do we see that the user will always have to close the PCI device to detach
   the hwpt? What happens if the user needs to connect to another hwpt or just
   needs another one?
2. Does it make sense to think about providing a way to detach the hwpt without
   closing the pci device?
3. Notice that this is more like a general comment regarding VFIO_DEVICE_ATTACH_IOMMUFD_PT than for this line

Reply:
I think it would be nice to provide that functionality. How to do that I am not sure. This series suggests to add an iommufd specific header file `vfn/iommu/iommufd.h` since this functionality is tied to iommufd and not generic like the rest of the API in `vfn/iommu.h`. In lieu of that, detaching could (should) be provided as well.

Reply:
Based on what we discussed offline, I would not implement a detach and leave it for when you have more information on how the library is being used.